src/PetriEngine/PQL/PredicateCheckers.cpp

```diff
@@ -69,116 +69,129 @@ namespace PetriEngine::PQL {
+            setConditionFound();
         }
     }
 
-    void IsReachabilityVisitor::_accept(const ACondition *element) {
+    void IsNotReachabilityVisitor::_accept(const ACondition *element) {
         if (!_is_nested) {
             if (auto cond = dynamic_cast<GCondition*>(element->getCond().get())) {
                 _is_nested = true;
-                cond->visit(*this);
+                cond->getCond()->visit(*this);
+                _is_nested = false;
+            } else {
+                setConditionFound();
             }
+        } else {
+            setConditionFound();
         }
     }
 
-    void IsReachabilityVisitor::_accept(const LogicalCondition *element) {
+    void IsNotReachabilityVisitor::_accept(const LogicalCondition *element) {
         if (_is_nested) {
             for(auto& c : element->getOperands())
             {
-                // This breaks the ANY pattern, we need to check that all operands are reachability
-                _condition_found = false;
                 c->visit(*this);
 
-                if(!_condition_found)
+                if(_condition_found)
                     break;
             }
+        } else {
+            setConditionFound();
         }
     }
 
-    void IsReachabilityVisitor::_accept(const CompareCondition *element) {
-        if (_is_nested) setConditionFound();
+    void IsNotReachabilityVisitor::_accept(const CompareCondition *element) {
+        if (!_is_nested) setConditionFound();
     }
 
-    void IsReachabilityVisitor::_accept(const NotCondition *element) {
+    void IsNotReachabilityVisitor::_accept(const NotCondition *element) {
         element->getCond()->visit(*this);
     }
 
-    void IsReachabilityVisitor::_accept(const BooleanCondition *element) {
-        if (_is_nested) setConditionFound();
+    void IsNotReachabilityVisitor::_accept(const BooleanCondition *element) {
+        if (!_is_nested) setConditionFound();
     }
 
-    void IsReachabilityVisitor::_accept(const DeadlockCondition *element) {
-        if (_is_nested) setConditionFound();
+    void IsNotReachabilityVisitor::_accept(const DeadlockCondition *element) {
+        if (!_is_nested) setConditionFound();
     }
 
-    void IsReachabilityVisitor::_accept(const UnfoldedUpperBoundsCondition *element) {
-        if (_is_nested) setConditionFound();
+    void IsNotReachabilityVisitor::_accept(const UnfoldedUpperBoundsCondition *element) {
+        if (!_is_nested) setConditionFound();
     }
 
-    void IsReachabilityVisitor::_accept(const ShallowCondition *element) {
-        if (_is_nested) setConditionFound();
+    void IsNotReachabilityVisitor::_accept(const ShallowCondition *element) {
+        if (!_is_nested) setConditionFound();
     }
 
-    void IsReachabilityVisitor::_accept(const QuasiLivenessCondition *element) {
+    void IsNotReachabilityVisitor::_accept(const QuasiLivenessCondition *element) {
         if (element->getCompiled())
             element->getCompiled()->visit(*this);
```

Comment:
Let us be conservative here (and in LivenessCondition & StableMarkingCondition too), and setConditionFound() if the expression was not compiled.